src/WebJobs.Script.WebHost/Security/KeyManagement/KeyVaultSecretsRepository.cs

```diff
@@ -0,0 +1,298 @@
+            }
+            if (!tasks.Any())
+            {
+                return null;
+            }
+            await Task.WhenAll(tasks);
+
+            FunctionSecrets functionSecrets = new FunctionSecrets()
+            {
+                Keys = new List<Key>()
+            };
+
+            foreach (Task<SecretBundle> task in tasks)
+            {
+                SecretBundle item = task.Result;
+                functionSecrets.Keys.Add(SecretBundleToKey(item, prefix));
+            }
+
+            return functionSecrets;
+        }
+
+        private string GetVaultBaseUrl()
+        {
+            return $"https://{_vaultName}.vault.azure.net";
+        }
+
+        public static Dictionary<string, string> GetDictionaryFromScriptSecrets(ScriptSecrets secrets, string functionName)
+        {
+            Dictionary<string, string> dic = new Dictionary<string, string>();
+            if (secrets is HostSecrets)
+            {
+                HostSecrets hostSecrets = secrets as HostSecrets;
+                if (hostSecrets.MasterKey != null)
+                {
+                    dic.Add(masterKey + "master", hostSecrets.MasterKey.Value);
+                }
+
+                if (hostSecrets.FunctionKeys != null)
+                {
+                    foreach (Key key in hostSecrets.FunctionKeys)
+                    {
+                        dic.Add($"{functionKeyPrefix}{Normalize(key.Name)}", key.Value);
+                    }
+                }
+
+                if (hostSecrets.SystemKeys != null)
+                {
+                    foreach (Key key in hostSecrets.SystemKeys)
+                    {
+                        dic.Add($"{systemKeyPrefix}{Normalize(key.Name)}", key.Value);
+                    }
+                }
+            }
+            else
+            {
+                FunctionSecrets functionSecrets = secrets as FunctionSecrets;
+                if (functionSecrets.Keys != null)
+                {
+                    foreach (Key key in functionSecrets.Keys)
+                    {
+                        dic.Add($"{functionPrefix}{Normalize(functionName)}--{Normalize(key.Name)}", key.Value);
+                    }
+                }
+            }
+            return dic;
+        }
+
+        public static string Normalize(string name)
+        {
+            string result = string.Empty;
+            foreach (char c in name)
+            {
+                if (char.IsLetterOrDigit(c))
+                {
+                    result += c;
+                }
+                else
+                {
+                    string charValue = ((int)c).ToString("D3");
+                    result += normalizedCharMark + charValue;
+                }
+            }
+            return result;
+        }
+
+        public static string Denormalize(string normalizedName)
+        {
+            string result = string.Empty;
+            for (int index = 0; index < normalizedName.Length; index++)
```

Comment:
This logic can probably be simplified with a regex replace call that parses the match value into the character

Reply:
I tried but it looks to hard. I do not want to spend lot of time on it.